### PR TITLE
feat: support for AArch64

### PR DESCRIPTION
**Pull request by RaitoBezarius**
This introduces a basic support of AArch64 through the introduction of System structures to write the correct filenames based on UEFI "specifications".

### CODE REVIEW

**Comment by RaitoBezarius** [comment]
https://github.com/NixOS/nixpkgs/pull/207171 will make it possible to test.

**Comment by RaitoBezarius** [comment]
```
cargo-nightly-aarch64-unknown-linux-gnu.tar.xz> trying https://static.rust-lang.org/dist/2022-11-21/cargo-nightly-aarch64-unknown-linux-gnu.tar.xz
cargo-nightly-aarch64-unknown-linux-gnu.tar.xz>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
cargo-nightly-aarch64-unknown-linux-gnu.tar.xz>                                  Dload  Upload   Total   Spent    Left  Speed
cargo-nightly-aarch64-unknown-linux-gnu.tar.xz> 100 6431k  100 6431k    0     0  1688k      0  0:00:03  0:00:03 --:--:-- 1707k
cargo> unpacking sources
cargo> unpacking source archive /nix/store/2s62393p2lmh8c5hh16j4ik2q9zhldk4-cargo-nightly-aarch64-unknown-linux-gnu.tar.xz
cargo> source root is cargo-nightly-aarch64-unknown-linux-gnu
cargo> setting SOURCE_DATE_EPOCH to timestamp 1668973825 of file cargo-nightly-aarch64-unknown-linux-gnu/version
cargo> patching sources
error: a 'x86_64-linux' with features {} is required to build '/nix/store/pn0cwp6jhmm5n28r7hrrwrw0mlrc24l5-append-initrd-secrets.drv', but I am a 'aarch64-linux' with features {benchmark, big-parallel, kvm, nixos-test}
```

**Comment by nikstur** [comment]
We (@RaitoBezarius & @nikstur) have agreed to only merge this once we have decent testing hardware (also for CI)

File: rust/lanzatool/src/esp.rs
Comment (blitz) [inline rust/lanzatool/src/esp.rs:14]: Yes. :laughing: 

Maybe we can just piggy back on [ProcessorArch](https://docs.rs/uefi/0.18.0/uefi/proto/debug/struct.ProcessorArch.html). This would avoid the awkwardly named `EfiFallback`.

Then we can just have some freestanding functions:

- `system_double_to_efi_arch(&str) -> ProcessorArch`
- `efi_fallback_filename(ProcessorArch) -> &Path`
- `systemd_filename(ProcessorArch) -> &Path`

PS. Technically a 32-bit UEFI system can host a 64-bit operating system. This happened on old Atom tablets... So technically a `x86_64-linux` system could be either 32 or 64-bit UEFI. In reality though, 32-bit UEFI should be pretty dead on x86.

File: rust/lanzatool/src/esp.rs
Comment (blitz) [inline rust/lanzatool/src/esp.rs:63]: Do we support "cross installation"? I.e. installing AArch64 from x86? If so, we should also use the right lanzaboote binary depending on the target system.

**Comment by blitz** [comment]
> We (@RaitoBezarius & @nikstur) have agreed to only merge this once we have decent testing hardware (also for CI)

Would testing in Qemu be good enough for now?

**Comment by RaitoBezarius** [comment]
> > We (@RaitoBezarius & @nikstur) have agreed to only merge this once we have decent testing hardware (also for CI)
> 
> Would testing in Qemu be good enough for now?

I am interested if you succeed, testing this in QEMU would require some sort of nested virtualization if you are on x86.

Maybe, if you have a good aarch64-linux machine, that would be nice, but the nix-community builder is a free Oracle VM and seems to choke somewhere in the VM startup.
My Pine64 A64 is too slow for this.

We would need a baremetal aarch64-linux to get this I suppose or someone with a PineBook Pro (or similar) AFAIK.

File: rust/lanzatool/src/esp.rs
Comment (RaitoBezarius) [reply rust/lanzatool/src/esp.rs:63]: AFAIK, here's what we have (with this PR):

- Cross builds: you can build aarch64 images with SecureBoot files using this, because `nixpkgs.hostSystem` / `nixpkgs.crossSystem` is automatically doing the hard work for us. We need to check if *cross compilation* works though, but *emulated compilation* do work (tested with `binfmt`).
- Cross installation: you are on an x86 machine, you want to install a AArch64 machine from there, because your processor can become an AArch64 after reboot, this, we do not really support AFAIK? Except for system hard disks being moved around, I'm not sure, this is a classical scenario we want to support in priority.

File: rust/lanzatool/src/esp.rs
Comment (RaitoBezarius) [reply rust/lanzatool/src/esp.rs:14]: I have noticed the 32-bits UEFI system thing, but I am… tempted to not support it at all, except if someone come with a PR and a test.

Agreed on the design.

Comment (nikstur) [reply rust/lanzatool/src/esp.rs:14]: Maybe we should just stick with the nix terminology and call it `System`? Although I personally hate the term "System", but maybe we should do it for consistency's sake. 

This could also be implemented upstream in bootspec, so the system field is more strongly typed. We could then just implement a trait for `System` like `impl SystemdBootPaths for bootspec::System`

Comment (blitz) [reply rust/lanzatool/src/esp.rs:14]: `System` is good. I don't see why we would need traits, but ymmv.

Comment (RaitoBezarius) [reply rust/lanzatool/src/esp.rs:14]: I guess it should be `systemd_filename(&self) -> Option<&Path>` because I'm not sure sd-boot supports all platforms, same goes for EFI fallback filenames. (does SPARC64 have UEFI support?)

Comment (nikstur) [reply rust/lanzatool/src/esp.rs:14]: > `System` is good. I don't see why we would need traits, but ymmv.

We would only need a trait if we want to implement something on a foreign type (i.e. a `System` enum from the bootspec lib).

**Comment by blitz** [comment]
I've converted this to draft to avoid people stumbling here and expecting everything to work. :) This needs some work to rebase etc.

**Comment by RaitoBezarius** [comment]
I will try to cleanup this PR so we can have AArch64.

File: rust/lanzatool/src/esp.rs
Comment (RaitoBezarius) [reply rust/lanzatool/src/esp.rs]: I implemented what we said, do not hesitate to reopen if not satisfying.

File: rust/lanzatool/src/esp.rs
Comment (RaitoBezarius) [reply rust/lanzatool/src/esp.rs]: Now I understand better your comment, we could support `targetPlatform` if we want. I suggest to support this in the future.

**Comment by RaitoBezarius** [comment]
This PR is now cleaned up.

File: rust/tool/src/esp.rs
Comment (nikstur) [inline rust/tool/src/esp.rs:108]: Are we assuming that the `system` will change between generations? Otherwise this doesn't seem right. Maybe we should add a parameter to `lzbt` that tells it the system before reading any bootspecs.

Comment (RaitoBezarius) [reply rust/tool/src/esp.rs:108]: Otherwise, we could take this less correct approach of installing lanzaboote for the current system sure. I don't have a strong opinion, I find this implementation elegant because it does handle multi targets for free.

File: nix/packages/uefi-run.nix
Comment (nikstur) [inline nix/packages/uefi-run.nix:1]: Shouldn't it be `hostPlatform`? Also why are we building the derivation ourselves? Can't we use the one from nixpkgs?

File: rust/tool/src/esp.rs
Comment (nikstur) [inline rust/tool/src/esp.rs:12]: I would prefer if this was `Arch` or `Architecture` instead of `System` because (a) `System` is a stupid name for it which nobody but Nix uses and (b) this enum really only contains the architecture part of the NixOS system double.

File: rust/tool/src/esp.rs
Comment (nikstur) [inline rust/tool/src/esp.rs:40]: Instead of `FromStr` we could just implement a `from_nixos_system` function. This way it is clear that we are constructing an architecture single from the nixos system double.

File: rust/tool/src/install.rs
Comment (nikstur) [inline rust/tool/src/install.rs:319]: Calling this function here means that for every generation, the systemd-boot binary on the ESP AND in the Nix Store is at least opened and checked. This seems pretty wasteful :D

Also, if we allow/treat the cursed case of generations with different architectures/systems, we cannot write the path to the systemd-boot binary into the efivars anymore.

All in all I'm more and more convinced that we should not allow/treat generations with different architectures. We should add system as a parameter to `lzbt`:

Comment (RaitoBezarius) [reply rust/tool/src/install.rs:319]: Good point; let's hard fail on multiple systems in generations and write the unique systemd boot once based on the unique value of system, ok for you?

Comment (RaitoBezarius) [reply rust/tool/src/install.rs:319]: (actually we don't care that much about efivars because we have EFI fallback dirs technically)

File: rust/tool/tests/common/mod.rs
Comment (nikstur) [inline rust/tool/tests/common/mod.rs:104]: The way you implemented this, I think, e don't really need the parameter here. We can just call `target_system_double()` from within `setup_toplevel`.

Comment (RaitoBezarius) [reply rust/tool/tests/common/mod.rs:104]: I wanted to ensure that everyone had the same target system double if we decided to pass mixed systems

File: rust/tool/src/install.rs
Comment (nikstur) [reply rust/tool/src/install.rs:319]: > Good point; let's hard fail on multiple systems in generations and write the unique systemd boot once based on the unique value of system, ok for you?

So you still want to read the system from the bootspec?

Comment (RaitoBezarius) [reply rust/tool/src/install.rs:319]: Yes because they are the real source of truth IMHO, if they have anomalous data, it's a problem we should report and we should not act based on external overrides from command line (yet?) IMHO.

File: rust/tool/tests/common/mod.rs
Comment (nikstur) [reply rust/tool/tests/common/mod.rs:104]: I'd say we reduce the noise now (call `target_system_double()`  from within `setup_toplevel()`) and then we can still add it as a parameter when we actually need/use it.

Comment (RaitoBezarius) [reply rust/tool/tests/common/mod.rs:104]: Agreed!

File: rust/tool/src/install.rs
Comment (nikstur) [reply rust/tool/src/install.rs:319]: I wouldn't do it because (a) this would complicate the code quite a bit, (b) we already read systemd from the command line  so we might as well do it for `system` (this would be effectively what systemd-boot-builder.py does but systemd-boot-builder.py relies on systemd being built for the right target), and (c) this would provide little benefit to the user.

Comment (RaitoBezarius) [reply rust/tool/src/install.rs:319]: Fair points. Let's go with your suggestion for now.

Comment (RaitoBezarius) [reply rust/tool/src/install.rs]: Done

**Comment by RaitoBezarius** [comment]
This is now ready again.

**Comment by RaitoBezarius** [comment]
I actually tested Lanzaboote on AArch64 with a lot of pain via manual QEMU stuff, this PR makes it possible on the long term to use Lanzaboote on AArch64 machines properly.

**Comment by RaitoBezarius** [comment]
@nikstur @blitz could I get approval? Hercules CI is not very useful because we don't have the proper devices for it, but also the PR bitrots :'( — I tested it with ALLOW_BROKEN, the compiler-rt is working in fact. I will push a PR upstream for that.

File: rust/tool/src/esp.rs
Comment (nikstur) [inline rust/tool/src/esp.rs:12]: This struct encodes a lot of information from differnte domains. We should make this more generic, move it out of the `esp.rs` module and with only one function besides the constructor: `efi_representation()` (or another appropriate name).  This function would be similar to https://github.com/systemd/systemd/blob/main/meson.build#L2114  The caller (i.e. in the esp module) then has the domain kowledge to assemble the filenames just from this information.

File: nix/modules/lanzaboote.nix
Comment (nikstur) [inline nix/modules/lanzaboote.nix:107]:
```suggestion
          --target-system ${config.nixpkgs.hostPlatform.system} \
```

File: rust/tool/src/install.rs
Comment (nikstur) [inline rust/tool/src/install.rs:92]: I'm not sure if this would change behaviour. Probably better to leave it as it was.

File: rust/tool/src/install.rs
Comment (nikstur) [inline rust/tool/src/install.rs:339]: This is where it shows that information is leaking when `Architecture` is in `esp.rs`. This path has nothing to do with the esp.

File: rust/tool/tests/common/mod.rs
Comment (nikstur) [inline rust/tool/tests/common/mod.rs:25]: This should be a conditionally compiled constant instead of a function.

File: rust/tool/tests/common/mod.rs
Comment (nikstur) [inline rust/tool/tests/common/mod.rs:138]: This shouldn't be a parameter since we set the system globally. With this parameter the diff would also be considerably smaller.

File: rust/tool/src/install.rs
Comment (nikstur) [inline rust/tool/src/install.rs:31]:
```suggestion
    architecture: Architecture,
```

Let's drop the `target_`. Cross compiling (terminology) is hard enough. As far as it concerns the installer, host, build, and target platform are all the same.

File: rust/tool/src/cli.rs
Comment (nikstur) [inline rust/tool/src/cli.rs:36]:
```suggestion
    system: String,
```

Same as with `architecture`. Any cross compilation is handled beforehand within Nix/NixOS.

File: rust/tool/src/install.rs
Comment (RaitoBezarius) [reply rust/tool/src/install.rs:31]: Maybe I got something wrong, but installer can install on a build != host != target platform.
Here, host is the UEFI platform of the target.

Comment (RaitoBezarius) [reply rust/tool/src/install.rs:31]: (except if we have lzbt touching efivars which is another problem.)

File: rust/tool/src/install.rs
Comment (RaitoBezarius) [reply rust/tool/src/install.rs:339]: Completely agreed.

File: rust/tool/src/install.rs
Comment (nikstur) [reply rust/tool/src/install.rs:31]: I don't think the installer can or should install from one architecture to another (e.g. from aarch64 -> x86). The only case where this would happen is when remote installing (of which I'm not really convinced).

That the host platform for the binaries is UEFI (and that these binaries are usually cross-compiled from x86 or aarch64) doesn't really matter for the installer. It only matters for the derivation that builds the stub.

Comment (RaitoBezarius) [reply rust/tool/src/install.rs:31]: While all of this might be true today, the installer is run as part of the switch-to-configuration phase, going forward if the day where you can "pre-switch-to-configuration" nixos systems, every install bootloader phase will have to deal with potentially build != host != target platforms.

I'm okay with assuming for now that the installer cannot deal with this scenario for now, but I don't believe it's a far-fetched thing and it's more correct.

> That the host platform for the binaries is UEFI (and that these binaries are usually cross-compiled from x86 or aarch64) doesn't really matter for the installer. It only matters for the derivation that builds the stub.

Yes, of course, it's just a correctness thing so we avoid being confused in the future if someone come up with unexpected usecases IMHO.

File: rust/tool/src/esp.rs
Comment (RaitoBezarius) [reply rust/tool/src/esp.rs]: This is now done, I used a Ext-style pattern for assembling the filenames based on the domain knowledge, is that fine with you?

File: nix/modules/lanzaboote.nix
Comment (RaitoBezarius) [reply nix/modules/lanzaboote.nix:107]: No, then, it should be `config.nixpkgs.targetPlatform.system` I believe.

Comment (RaitoBezarius) [reply nix/modules/lanzaboote.nix:107]: But this can create surprising behaviors IMHO.

File: rust/tool/src/install.rs
Comment (RaitoBezarius) [reply rust/tool/src/install.rs]: Right, it was a mistake.

File: rust/tool/src/install.rs
Comment (RaitoBezarius) [reply rust/tool/src/install.rs:340]: But then, are you okay with the ext trait solution? Should I move it into systemd.rs ?

File: rust/tool/tests/common/mod.rs
Comment (RaitoBezarius) [reply rust/tool/tests/common/mod.rs]: Fixed

File: rust/tool/src/cli.rs
Comment (RaitoBezarius) [reply rust/tool/src/cli.rs:36]: I don't like this; unprefixed systems are the reason for why we are always confused. I am okay with dropping target, but I would like to have an alternative: host or build.

File: rust/tool/tests/common/mod.rs
Comment (RaitoBezarius) [reply rust/tool/tests/common/mod.rs:26]: Done

File: rust/tool/src/cli.rs
Comment (nikstur) [reply rust/tool/src/cli.rs:36]: That's fair but then we should use the same terms as Nixpkgs, maybe just `hostPlatform`? But `hotsPlatform` doesn't have a string representation, right?

Comment (RaitoBezarius) [reply rust/tool/src/cli.rs:36]: hostPlatform has a string representation

**Comment by RaitoBezarius** [comment]
The PR is now cleaned up again and we went through details IRL with @nikstur.

**Review by nikstur — APPROVED** [review]

**Comment by RaitoBezarius** [comment]
Given we met again an issue with systemd bootaa64.efi not containing the right sections, we will hold off until `nixos-unstable-small` gets https://github.com/NixOS/nixpkgs/pull/243242.

**Comment by RaitoBezarius** [comment]
Given we tested this a lot, I am inclined to merge now before it bitrot again, we can move the testing to actual runtimes as our CI is meh for this in general.

File: flake.nix
Comment (SuperSandro2000) [inline flake.nix:82]:
```
       error: attribute 'hostPlatform' missing

       at /nix/store/8hqjsxybh8fvkij104linyyp3gi3c560-source/flake.nix:82:27:

           81|         let
           82|           rustTarget = "${pkgs.hostPlatform.qemuArch}-unknown-uefi";
             |                           ^
           83|           pkgs = import nixpkgs {
       Did you mean rustPlatform?
```

Fix is in #231

**Pull request merged**
